stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -1599,17 +1608,26 @@ extension ${Self} {
   /// can't be represented exactly, the result is `nil`.
   ///
   /// - Parameter value: The integer to represent as a floating-point value.
-%   if srcBits < SignificandBitCount:
+%   if srcBits <= dstBits:
   @available(*, message: "Converting ${That} to ${Self} will always succeed.")
 %   end
   @inlinable // FIXME(sil-serialize-all)
   @inline(__always)
   public init?(exactly value: ${That}) {
     _value = Builtin.${sign}itofp_${ThatBuiltinName}_FPIEEE${bits}(value._value)
 
-%   if srcBits >= SignificandBitCount:
-    guard let roundTrip = ${That}(exactly: self),
-      roundTrip == value else {
+%   if srcBits > dstBits:
+    // The magnitude of the integer must be representable by the significand.
+    // Omit trailing zeros that can be encoded by the exponent.
+    let absv = value.magnitude
+    let mask: ${That}.Magnitude = ~((1 << (${SignificandBitCount} + 1)) - 1)
```

Comment:
I'm concerned as to what happens here if `srcBits == SignificandBitCount` (assuming you meant to preserve the original condition `srcBits >= SignificandBitCount` modulo refactoring). It appears this calculation would not be correct because it'd overflow. Rather, I think you'd want to construct the mask by right-shifting `${That}.Magnitude.max` the appropriate number of bits. (Note that we now have the guarantee that `Magnitude` is unsigned as of a recent commit.)

Reply:
The check above excludes the `srcBits == SignificandBitCount` case, so overflow can't occur here.